Add initial e2e testing infrastructure

marsella opened the pull request:
This adds the basic testing harness to run our end-to-end tests. It makes two large sweeps of changes:

## Restructures zkchannels architecture
Before, the zkchannels protocol logic was encoded as a binary crate, which used several libraries. Now, the bulk of the protocol logic has been moved to the zkchannel library, which releases a public API to call `run` on each command line operation. 

The `Show` and `List` command line operations were also modified to return content (e.g., details about a particular channel), rather than printing it to stdout. To support this, I added a `PublicChannelDetails` struct for each party, which encodes the reasonable details about a channel to release to the user.

## Creates testing harness
Creates a custom integration test harness (following [the advice of a blog post](https://www.infinyon.com/blog/2021/04/rust-custom-test-harness/) to configure and run a customer watcher and merchant server, execute all tests, and close down the long-running tasks. 

Tests are specified as a set of operations and their expected outcome. If the servers do not reflect the expected outcome after an operation completes, the test fails. The expected outcome currently consists of the channel status according to each customer and an indicator that a process threw an error. This can be expanded as additional operations with more complex features are added (see, e.g. #350, #352).

I monitor errors thrown by the servers by writing them to a file and tagging them (via `tracing`) with _which_ server created them. The current implementation deletes the log after each test (to avoid getting mixed up with errors thrown by a previous test). I hard-coded a pause (see `wait_time`) after certain operations to make sure the servers have enough time to finish their business. This seems like it could be a pain point in the future, but it works for the current (limited) set of tests.

This is currently hard-coded for the Tezos specification. Note that the Tezos integration cannot support parallelized operations; these tests are run serially by default.

## Specific points for reviewers
I made a bunch of new issues (#350 - 360, collected under #69) to describe the remaining work necessary to get a fully working test harness. If some of these should be incorporated into this PR, please note in review.

I originally made a main and a common file that I expected to be separated roughly along Tezos/non-Tezos lines. I am not sure the division I ended up with makes sense any more. Any thoughts on organization of these? I am considering just sticking everything into one massive file and trying to make a reasonable abstraction later, when we have something to abstract over. It may also be more reasonable to put the test specs themselves in a separate file, because I think there will be a _lot_ of them.

The CI for pull requests does not run these tests by default. Should I try to find a way to manually trigger them (just to prove this works)? 

Closes #133 
Closes #345 - this is no longer necessary
Addresses #69, #338 but does not complete them.
Addresses some tasks in #360

marsella commented:
Hmm, I thought the change to the yml file would _not_ run these tests, but apparently it does. I will look into fixing that (they failed because the server infrastructure is not set up).

gijsvl commented on integration_tests/common.rs:
```diff
@@ -0,0 +1,261 @@
+            ("certificate", "localhost.crt"),
+        ])
+        .into_iter()
+        .fold("\n[[service]]".to_string(), |acc, (key, value)| {
```
Isn't this printing [[service]] between every line?

marsella replied:
Nope, the first argument to `fold` is the initial value for the accumulator (`acc`). So the first time the second argument function is called, its arguments are `("[[service]]", ("address", "localhost"))`. Then the next time, the arguments are `("[[service]]\n address = localhost", ("private_key", "localhost.key")` and so on.

gijsvl commented on integration_tests/common.rs:
```diff
@@ -0,0 +1,261 @@
+
+    Ok(logs
+        .lines()
+        .filter(|s| s.contains("zeekoe::"))
```
This is okay for me, I just think you can do the filtering of log messages using the correct subscriber, but it might be less flexible than this, so this is probably better.

marsella replied:
No, I just did this because I was scared to set up the `EnvFilter` but I will fix it! We can always expand the filter if we want to look for errors in other crates.

marsella replied:
I fell down a rabbit hole but I adding an `EnvFilter` seems to wipe out the per-party tagging that I did with `instrument`. I am not sure why at this point, but if I can't have them simultaneously I will ditch the filter.

gijsvl commented on integration_tests/main.rs:
```diff
@@ -0,0 +1,276 @@
+        .expect("Failed to load merchant config");
+
+    // Give the server some time to get set up
+    tokio::time::sleep(tokio::time::Duration::new(5, 0)).await;
```
You  made  a comment about this as well I believe, is this something we can do based upon the logging? I believe the server prints a certain message when it's ready...

marsella replied:
Yes! It required more code than I expected, but I added this functionality.

gijsvl commented on integration_tests/main.rs:
```diff
@@ -0,0 +1,276 @@
+            .await;
+
+            // Sleep until the servers have finished their thing, approximately
+            tokio::time::sleep(op.wait_time()).await;
```
Can we do pulling of logs (with timeout) instead?

marsella replied:
I can try this. My concerns are:
- There are a lot of various correct "done" scenarios. I'm hesitant about hard-coding a "done" message for every `Operation` into zeekoe, especially when there are multiple valid paths. I would be concerned about missing one.
- I'm not sure what would happen if we find an actual failure case (e.g., where the server does not get to the "done" message because something errored). We would still need a timeout larger than the max expected time for any of these things to run.

marsella replied:
I just added some infra in 2320d5c to wait for specific logs - in that case, to see if the merchant server and customer watcher are up. I can reuse a lot of that for this situation too.

marsella replied:
I'm getting stuck on order of operations and correctness conditions here. For example, I would need to check the merchant server log for a "successfully established" log. But if the channel was supposed to fail to establish, then I need to skip that. But I don't want to check whether it was supposed to fail until after I know the server is done. 

I think I am going to stick with the timeout approach for now and we can discuss how this could work in a call at some point.

gijsvl commented on src/cli/customer.rs:
```diff
@@ -26,7 +26,7 @@ pub struct Cli {
 pub enum Customer {
     List(List),
-    // Show(Show),
+    Show(Show),
```
Why was it commented out?

marsella replied:
It wasn't implemented before for a single channel. I think it was just lower priority because we never checked a single channel at a time.

gijsvl commented on src/cli/customer.rs:
```diff
@@ -49,8 +49,8 @@ pub struct List {
-    /// The channel ID, or a unique prefix of it.
-    pub prefix: String,
+    /// The channel label.
+    pub label: ChannelName,
```
Did this really change? (I mean in functionality), can we still do a prefix instead of full label?

marsella replied:
Previously, this wanted a prefix of the channel ID, the long hex string, rather than the human-readable channel name. We don't have a use for the ID query right now, so I decided to remove it.

gijsvl commented on src/zkchannels/merchant/manage.rs:
```diff
@@ -43,33 +88,23 @@ impl Command for List {
-                })
-                .to_string()
-            );
+            Ok(serde_json::to_string(&PublicChannelDetails::from(details))?)
```
I love it, this is so much better than the printing!

marsella replied:
Yay! I just propagated this change to both `List` commands as well, for consistency.

gijsvl commented on .github/workflows/e2e-test.yml:
```diff
@@ -59,10 +60,7 @@ jobs:
-            nohup python3 test-zeekoe.py merch-setup --url "http://my-sandbox:20000" -v &
-            nohup python3 test-zeekoe.py cust-setup --url "http://my-sandbox:20000" -v &
-            python3 test-zeekoe.py test-all
+            cargo test --all-features --all-targets --verbose
```
Can we do something with the command so it only runs the integration tests, it runs unit tests as well?

marsella replied:
Good question. We can do
```
cargo test --all-features --tests integration_tests
```
Then we can change the pull request yml to do
```
cargo test --all-features --lib --bins --examples --benches
```
However, these two commands would exclude any new integration tests we add (e.g. in `test/`). Is this worth being concerned about? Presumably if we write new tests, we will pay attention to whether they're actually getting run.

I have not found a way to tell cargo to ignore / skip `integration_tests` but run everything else.

gijsvl commented on .github/workflows/rust.yml:
```diff
@@ -41,8 +41,8 @@ jobs:
-      - name: Run all tests except doctests
-        run: cargo test --all-features --all-targets --verbose
+      - name: Run all unit + library tests (not doc or integration tests)
+        run: cargo test --all-features --all-targets --lib --verbose 
```
This seems to not exclude the integration tests, but I guess you figured given your last comment on this PR.

marsella replied:
Fixed this, see note below. I needed to remove `--all-targets` and replace it with `--lib --bins`, cargo was choosing the most expansive option.

gijsvl reviewed (changes requested):
This is very impressive coding, I've said it before, this is a very nice setup. Thanks a lot for creating this!
I've added some minor comments. I wasn't able to run it locally (somehow python lost libsodium, but only in the test, not when I just do a manual setup). It seems that there is no run during the PR of the e2e tests, maybe we should activate it temporarily for this PR and remove before merge?

gijsvl commented:
Although the e2e github action seems to succeed, there is not much evidence about the tests, it would be good if there was some output about the success of the tests. Similar to regular tests.

marsella commented on integration_tests/common.rs:
```diff
@@ -0,0 +1,261 @@
+        ("database", "{ sqlite = \"customer.db\" }"),
+        ("trust_certificate", "\"localhost.crt\""),
+        ("tezos_account", "{ alias = \"alice\" }"),
+        ("tezos_uri", "\"http://localhost:20000\""),
```
Good call, I forgot about that. Added #362

marsella commented:
Thanks for triggering them correctly. I think I have fixed the config issues and printed output correctly now. Also, I updated the integration test crate to cleanly return an error code instead of panicking.

For reference, I leaned that there are 5 different test targets (`--lib`, `--bins`, `--tests`, `--examples`, and `--benches`). Specifying `--all-targets` runs all of them. The behavior we want is to run library and binary tests at pull, and everything in the e2e tests. Before, I was incorrectly specifying `--all-targets --lib` at pull, which just ran everything instead of only library tests.

Will we have to update the yml file before merging to un-specify e2e tests at pull request?

marsella commented:
Hmm, the changes in 37615c2 didn't fix the output like I thought they would. I get a bunch of info when I run it locally. I am not actually sure whether it is running `integration_tests` right now, it doesn't even have the line `Running integration_tests/main.rs` that I expected to see.

marsella commented on integration_tests/common.rs:
```diff
@@ -0,0 +1,261 @@
+    let _ = fs::create_dir("integration_tests/gen");
+
+    // Create self-signed SSL certificate config file and certificate
+    File::create("integration_tests/gen/ssl_config")
```
Yes, we definitely should. I thought this would be annoying to do, but it was not. Fixed now.

marsella commented on Dockerfile:
```diff
@@ -30,7 +30,7 @@ RUN git clone https://github.com/boltlabs-inc/zeekoe.git
 
 RUN git submodule update --init --recursive
-RUN ./dev/generate-certificates; CARGO_NET_GIT_FETCH_WITH_CLI=true cargo build --features "allow_explicit_certificate_trust"
+RUN ./dev/generate-certificates; CARGO_NET_GIT_FETCH_WITH_CLI=true cargo build --features "allow_explicit_certificate_trust allow_custom_self_delay"
```
long story short - no. This makes sure that we allow a short self-delay so the tests actually finish in less than 2 days.

marsella commented on integration_tests/common.rs:
```diff
@@ -0,0 +1,303 @@
+/// The minimum required depth to originate contracts on the Tezos blockchain.
+static MINIMUM_LEVEL: u64 = 120;
+
+pub async fn await_leveled_blockchain(
```
This takes some time but not too much, especially now that you fixed the ordering of jobs.

marsella commented on .github/workflows/e2e-test.yml:
```diff
@@ -49,7 +52,8 @@ jobs:
         run: docker run --rm --name my-sandbox --detach -p 20000:20000 tqtezos/flextesa:20210602 flobox start
 
-      - uses: addnab/docker-run-action@v3
+      - name: Configure tezos client and run all test targets
```
Fixed.

gijsvl reviewed (approved):
Looks very good to me, the caching on the build also works smoothly (although the sandbox hasn't finished mining enough blocks now ;-)). Really nicely done, this was a tough one, but totally worth it. Thank you so much for all your time on this!
